### PR TITLE
Fix shadows not rendering with orthographic render

### LINKED ISSUE #21: AerialPerspective: Shadows not rendering with orthographic camera
You might have mentioned this in #15 but I don't think I fully understood. When working on adding the toggle to the demo page I'm seeing that the scene doesn't seem to render shadows. See the same view before and after ortho toggle. Note the lack of shadows on the mountain: 

image 
 image

And the normals before and after switching (from a different area, of course 😅):

image image

It looks like this is due to the `correctGeometricError` function that interpolates normals as the camera zooms out, calculating the interpolation factor based on camera height. I'll look into some other ways of addressing this for an orthographic camera.

### PR DESCRIPTION
**Pull request by gkjohnson**
Fix #21 

Instead of using the height to determine which set of normals to use and interpolate we use the projected size of the globe at the surface of the ellipsoid. The min and max interpolation factors were chosen to roughly match the two heights / transition points that were previously chosen with the existing perspective camera fov.

Another benefit of this is that these settings will automatically account for different perspective camera FoVs in addition to an orthographic projection. With this change the `geometricErrorAltitudeRange` is no longer needed unless we want to expose the min / max parameters as variables.

_Also_ this interpolation value can be computed on the CPU and passed as a uniform which would simplify some of the shader calculations. If you'd like me to move the calculation I can do that, as well!

Perspective / Orthographic Shadows

image image

### CODE REVIEW

**Comment by shotamatsuda** [comment]
Thanks you! This is really insightful.
Please give me some time to decide whether to expose the interpolation edges of the projected radius. I had exposed `geometricErrorAltitudeRange` because I thought it depended on the the target screen space error.

> Also this interpolation value can be computed on the CPU and passed as a uniform which would simplify some of the shader calculations.

I agree. I’ll move it outside of the shader.

**Comment by gkjohnson** [comment]
> Please give me some time to decide whether to expose the interpolation edges of the projected radius. I had exposed geometricErrorAltitudeRange because I thought it depended on the target screen space error.

My typical feeling is to expose fewer options until there's a reason to since I think it's best to keep things simple for users and this is hard to explain. Since this should "just work" with multiple camera types and fields of view I think it's okay to remove. But of course this is just my opinion!

There is one last notable artifact when performing the camera transition where the atmosphere overlay color "pops" when switching to the orthographic camera. This is likely because we're computing the ray origin differently for the different camera types (ortho origin is at the near clip plane, perspective camera is the camera origin). I'll look into this, as well, but I'll have to think through whether it will cause floating point error problems and how to address it - slowly but surely!

File: packages/atmosphere/src/shaders/aerialPerspectiveEffect.frag [hunk 1]
```diff
@@ -20,18 +20,20 @@ vec3 readNormal(const vec2 uv) {
...
-  inout vec3 worldNormal
-) {
+void correctGeometricError(inout vec3 worldPosition, inout vec3 worldNormal) {
+  // calculate the projected scale of the globe in clip space
+  float maxRadius = max(ellipsoidRadii.x, max(ellipsoidRadii.y, ellipsoidRadii.z));
+  vec4 projectedPoint = projectionMatrix * vec4(0, maxRadius, - cameraHeight, 1);
+  projectedPoint /= projectedPoint.w;
+
+  // transition between the zoomed out and zoomed in normals based on projected size
+  float t = smoothstep(41.5, 13.8, projectedPoint.y);
```
Comment (gkjohnson) [inline packages/atmosphere/src/shaders/aerialPerspectiveEffect.frag:30]: One thing that might be worth noting is that these numbers can be thought of as "how many times larger is the projected size of the globe than the height of the viewport."

**Comment by gkjohnson** [comment]
> There is one last notable artifact when performing the camera transition where the atmosphere overlay color "pops" when switching to the orthographic camera.

It actually looks like this is caused by the "ellipsoid origin" being recalculated based on the camera height in the vertex shader (which would need to be changed to the same logic as is used in the fragment shader). Removing the condition doesn't seem to significantly impact the look of the scene, though, and it's not clear to me exactly what it's for. 

https://github.com/takram-design-engineering/three-geospatial/blob/ba442fde16cf183715a2d22d0d0ac628523f6f09/packages/atmosphere/src/shaders/aerialPerspectiveEffect.vert#L45-L54

**Comment by shotamatsuda** [comment]
> Removing the condition doesn't seem to significantly impact the look of the scene, though, and it's not clear to me exactly what it's for.

Because we're interpolating the ellipsoid into a true sphere in correctGeometricError, it is no longer valid to apply altitude correction, which is to address the differences between ellipsoid and sphere. Without this, the atmosphere becomes thinner atmosphere at the poles and thicker at the equator (the atmosphere's radius is set in-between).

Left: with this interpolation
Right: without it

 

The difference is very subtle, though.

**Comment by gkjohnson** [comment]
Okay! I've just adjust the PR so the interpolation factor is calculated on the GPU and uploaded as a uniform. The `geometricErrorAltitudeRange` uniform is still present in the `AerialPerspectiveEffect` class but I can remove if you decide to not expose it for now. Otherwise I can move the constants to that field.

I've also added a line to output color in the `stars.frag` file since I was getting the following gl warning without. I'm not sure why we have to write colors when every fragment is getting discard, though 🤷 

```
[.WebGL-0x120005c7a00] GL_INVALID_OPERATION: Active draw buffers with missing fragment shader outputs.
```

One other improvement may be to just remove the stars object from the scene entirely when an ortho camera is present so it's not rendered at all.

**Comment by shotamatsuda** [comment]
Looks fantastic! I’ve decided to remove the uniform until it’s truly necessary, as you said.

I wasn’t aware that some output is required when every fragment is discarded. Thanks for pointing that out!

I’m merging it now. Thank you so much for the improvement!

**Pull request merged**